pb/forms_test.go

```diff
@@ -0,0 +1,240 @@
+package merkledag_pb
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"testing"
+)
+
+func TestForm_empty(t *testing.T) {
+	n := &PBNode{}
+	expectedBytes := ""
+	expectedForm := `{
+	Data: nil,
+	Links: nil
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Data_zero(t *testing.T) {
+	n := &PBNode{Data: make([]byte, 0)}
+	expectedBytes := "0a00"
+	expectedForm := `{
+	Data: 0x,
+	Links: nil
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Data_some(t *testing.T) {
+	n := &PBNode{Data: []byte{0, 1, 2, 3, 4}}
+	expectedBytes := "0a050001020304"
+	expectedForm := `{
+	Data: 0x0001020304,
+	Links: nil
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+// a zero-length links array round-trips as nil
+func TestForm_Links_zero(t *testing.T) {
+	n := &PBNode{Links: make([]*PBLink, 0)}
+	expectedBytes := ""
+	expectedForm := `{
+	Data: nil,
+	Links: nil
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Data_some_Links_zero(t *testing.T) {
+	n := &PBNode{Data: []byte{0, 1, 2, 3, 4}, Links: make([]*PBLink, 0)}
+	expectedBytes := "0a050001020304"
+	expectedForm := `{
+	Data: 0x0001020304,
+	Links: nil
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Links_empty(t *testing.T) {
+	l := &PBLink{}
+	n := &PBNode{Links: []*PBLink{l}}
```

Comment:
what about `&PBNode{Links: []*PBLink{nil}}`? would that be valid? It certainly is in plain Go, and seems different from the other cases you have here.

Reply:
My Go isn't strong enough to decipher that one. Is it that there's 1 element and it's `nil`?

Here's what happens:

```
--- FAIL: TestForm_Links_emptynull (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x67a626]
```

Comes from this:

```go
    for iNdEx := len(m.Links) - 1; iNdEx >= 0; iNdEx-- {                        
      {    
        size, err := m.Links[iNdEx].MarshalToSizedBuffer(dAtA[:i])                            
```

trying to access properties on an element of the `Links` array inside `PBLink#MarshalToSizedBuffer()` throws the segfault.

I don't think I want to start actually changing things outside of these tests. In JS I'm asserting that every element in `Links` is an object, which at least covers the `&PBLink{}` case and will reject the `{nil}` case.

Reply:
> Is it that there's 1 element and it's `nil`?

Yep.

Sounds good if it's not interesting. I only flagged it because it was the only edge case you seemed to be missing on the Go side.